README.md

```diff
@@ -78,30 +88,45 @@ Traces can be written as lists of 4-tuples:
 ]
 ```
 
-Indexed ZSets should be merged into a trace using the key, value, timestamp, and weight associated with each element. ZSets should be merged into the trace by treating the element itself as a key, and associating no value with the delta.
+Indexed ZSets SHOULD be merged into a trace using the key, value, timestamp, and weight associated with each element.
+
+ZSets SHOULD be merged into the trace by treating the element itself as a key, and associating no value with the delta.
 
 Implementations MUST support efficient sequential access of traces, first by key, and then by value, and lastly by timestamp.
 
 ## 2.4 Weight
 
-An integer associated with each element of a ZSet. Positive weights indicate the number of derivations of that element within the ZSet, and negative weights indicate the number of deletions of that element.
+Weights describe the number of downstream dependency count for a [ZSet]. It MUST be represented by an integer, and MAY be either positive or negative.
 
-## 2.5 Time
+Positive weights indicate the number of derivations of that element within the ZSet. Negative weights indicate the number of deletions of that element.
 
-PomoFlow represents [timestamp]s for the root circuit using a counter which denotes the current epoch.
+## 2.6 Time
 
-Every recursive subcircuit refines its parent's timestamp using a pair which further associates that timestamp with the iteration count through the subcircuit. These pairs are ordered using product order.
+Dataflow engines require a concept of ordering. Each node in the circuit MUST be labelled with an incrementing integer, called its "epoch". PomoFlow represents [timestamp]s for the root circuit using a counter which denotes the current epoch.
```

Comment:
I think there might be some confusion somewhere, because the nodes aren't labelled with their epoch: "epoch" is just the name given to the root circuit's timestamp

Reply:
Oooh I also missed some references to PomoFlow in here. kk fixing thanks

Reply:
To save you a click, here's the update:

> Dataflow engines require a concept of ordering. Each node in the circuit MUST be labeled with an incrementing integer that represents the number of times that it has gone through that subcircuit. [DBSP] represents [timestamp]s for the root circuit using a counter which denotes the current epoch.

Reply:
That seems mostly right, but it isn't necessarily that each node needs to be labelled than that each operation in the circuit maintains its own internal clock. To me, labeling the nodes reads like a static operation that happens once, at graph construction time.

Reply:
Ah indeed: badly phrased. How about "each node in the circuit maintaines a counter, or 'clock'"?

When does the internal counter update? I'm skimming the paper, and I'm mainly getting brief references to the counter but no clear rules:

<img width="425" alt="Screenshot 2023-05-23 at 21 34 39" src="https://github.com/RhizomeDB/dbsp-notes/assets/1052016/c28482eb-583a-456a-bb41-96eb67ef44a0">

Does it increment every time it's touched, or is there some more complex algorithm?

Reply:
Yeah about that 😅

For the most part, it increments after every tuple it processes, and that's probably a good baseline behavior to describe here. Some of the more complex operators break that rule a little bit, as part of performance optimizations I didn't get into here. I'd have to go back to my implementation to find the details though.

Generally though, each node has a clock_start and clock_stop hook that gets called before and after evaluating that node, and they handle updating the node's clock and doing any additional bookkeeping that needs to happen alongside that update.

It gets a bit more complex dealing with subcircuits, because nodes are also notified when any of their parent circuits start or stop a clock cycle, and so both my implementation and dbsp itself parameterize those hooks over a "scope" that gives the layer of the circuit the clock event corresponds to. That ends up being used to do things like clean up state that's no longer needed after an epoch finishes.

Reply:
> Dataflow engines require a concept of ordering. Each node in the circuit contains an incrementing integer "clock". In most (but not all) cases, this represents the number of times that the node has been touched. In general, each node has `clock_start` and `clock_stop` hooks that get called before and after evaluating that node.
>
> Subcircuits are more complex: nodes are also notified when any of their parent circuits start or stop a clock cycle. Implementations generally parameterize those hooks over a "scope" that gives the layer of the circuit the clock event corresponds to. This is useful for things like cleaning up state at the end of an epoch.

More or less?

Reply:
Looks great! I think we can fill in the details when we implement this in Rhizome.